### ISSUE TITLE
Problem with vagrant-shell script

### ISSUE THREAD

**Issue by johnSchnake** [kubernetes-sigs/sig-windows-dev-tools#21]
```
    winw1: Running: sync/kubejoin.ps1 as c:\tmp\vagrant-shell.ps1
    winw1: At C:\tmp\vagrant-shell.ps1:1 char:4
    winw1: +  --cri-socket="npipe:////./pipe/containerd-containerd"--cri-socket "n ...
    winw1: +    ~
    winw1: Missing expression after unary operator '--'.
    winw1: At C:\tmp\vagrant-shell.ps1:1 char:4
    winw1: +  --cri-socket="npipe:////./pipe/containerd-containerd"--cri-socket "n ...
    winw1: +    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    winw1: Unexpected token 'cri-socket="npipe:////./pipe/containerd-containerd"--cri-socket' in expression or statement.
    winw1:     + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    winw1:     + FullyQualifiedErrorId : MissingExpressionAfterOperator
```

I'm trying to get this to run on my Macbook but ran into #20 and also the above issue. Not certain yet how to resolve this one; I was hoping it would be apparent to someone more familiar with the codebase already. Otherwise I'm happy to dig into it a bit.

**Comment by FriedrichWilken** [comment]
John, thanks for looking into this and giving feedback. Are you on an Intel or M1 MacBook? @jayunit100  , did you run into similar problems on your M1?

**Comment by johnSchnake** [comment]
Intel; 2.2 GHz 6-Core Intel Core i7

**Comment by jayunit100** [comment]
never saw this !!!!!!!!!!

**Comment by jayunit100** [comment]
thats happening INSIDE of the powershell .  

- Is there a possibility the intermediate `join` file isnt being written out properly? 
- @FriedrichWilken can we add a thing to validate/templatize that intermediate file more robustly ?

**Comment by johnSchnake** [comment]
join.txt has
```
 --cri-socket="npipe:////./pipe/containerd-containerd"
```

prepjoin.ps1 creates the kubejoin.ps1 and results in:
```
 --cri-socket="npipe:////./pipe/containerd-containerd"--cri-socket "npipe:////./pipe/containerd-containerd"
```

which seems odd but explicit per the prepjoin code. Also seems like a space may be missing.

**Comment by FriedrichWilken** [comment]
I also found this in `master.sh` :
```echo "`cat /var/sync/join.ps1` --cri-socket=\"npipe:////./pipe/containerd-containerd\"" > /var/sync/join.txt```
Something is off.

**Comment by sladyn98** [comment]
It looks like `master.sh` puts :

```
 --cri-socket="npipe:////./pipe/containerd-containerd"
```

inside `join.txt ` and then `prepjoin.ps1` creates `kubejoin.ps` and again appends the above to it resulting in it getting added twice. Ideally this should happen only once. A simple fix could be making sure `kubejoin.ps1` is created correctly
@FriedrichWilken You could make the templating more robust, by getting rid of `prepjoin.ps1` responsiblity of creating the file and letting master.sh handle it or make a separate file that takes care of it

**Comment by FriedrichWilken** [comment]
Yes, I have to wrap my head around all the recent changes. I lost track.

**Comment by johnSchnake** [comment]
I tried just removing the part of master.sh where it added to that file; so it would just exist once.

Got
```
==> winw1: Running provisioner: shell...
    winw1: Running: sync/kubejoin.ps1 as c:\tmp\vagrant-shell.ps1
    winw1: At C:\tmp\vagrant-shell.ps1:1 char:4
    winw1: +  --cri-socket="npipe:////./pipe/containerd-containerd"
    winw1: +    ~
    winw1: Missing expression after unary operator '--'.
    winw1: At C:\tmp\vagrant-shell.ps1:1 char:4
    winw1: +  --cri-socket="npipe:////./pipe/containerd-containerd"
    winw1: +    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    winw1: Unexpected token 'cri-socket="npipe:////./pipe/containerd-containerd"' in expression or statement.
    winw1:     + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    winw1:     + FullyQualifiedErrorId : MissingExpressionAfterOperator
The following WinRM command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!
```

The file looks OK; just the single cri-socket flag but I'm assuming there should have been more to it. Will try and take a look at it soon.

**Comment by FriedrichWilken** [comment]
I get a similar error:
```
    winw1: At C:\tmp\vagrant-shell.ps1:2 char:3
    winw1: + --cri-socket="npipe:////./pipe/containerd-containerd"
    winw1: +   ~
    winw1: Missing expression after unary operator '--'.
    winw1: At C:\tmp\vagrant-shell.ps1:2 char:3
    winw1: + --cri-socket="npipe:////./pipe/containerd-containerd"
    winw1: +   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    winw1: Unexpected token 'cri-socket="npipe:////./pipe/containerd-containerd"' in expression or statement.
    winw1: At C:\tmp\vagrant-shell.ps1:3 char:3
    winw1: + --cri-socket "npipe:////./pipe/containerd-containerd"
    winw1: +   ~
    winw1: Missing expression after unary operator '--'.
    winw1: At C:\tmp\vagrant-shell.ps1:3 char:3
    winw1: + --cri-socket "npipe:////./pipe/containerd-containerd"
    winw1: +   ~~~~~~~~~~
    winw1: Unexpected token 'cri-socket' in expression or statement.
    winw1:     + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    winw1:     + FullyQualifiedErrorId : MissingExpressionAfterOperator
The following WinRM command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!
```

I have to catch up with all the recent changes to the project.

**Comment by FriedrichWilken** [comment]
I see now, 
1. `master.sh` writes the `kubeadm token create --print-join-command` to `join.txt`
2. `master.sh` appends `--cri-socket "npipe:////./pipe/containerd-containerd"`
3. `prejoin.ps1` takes `join.txt`, adds `--cri-socket "npipe:////./pipe/containerd-containerd"` AGAIN and creates a `kubejoin.ps1` with all the text
4. vagrant runs `kubejoin.ps1`

But I think the actual problem might just be the fact that there is a newline between the join command and `--cri-socket ...`.

So what I did is getting rid of `kubejoin.ps1`, `join.txt` and `prejoin.ps1`. Then I created an empty `join.ps1` (vagrant will complain if scripts are missing). The 
1. `master.sh` deletes the "old" `join.ps1`
2. `master.sh` creates a new `join.ps1` with the fresh join command
3. `master.sh` replaces (via `sed`) "--discovery-token-ca-cert-hash"  with 
    "--cri-socket="npipe:////./pipe/containerd-containerd" --discovery-token-ca-cert-hash"
Using `sed` might be a little cumbersome, but it seems to do the job.

Now I get a new error:
`==> winw1: Running provisioner: shell...
    winw1: Running: sync/join.ps1 as C:\tmp\vagrant-shell.ps1
    winw1: [preflight] Running pre-flight checks
    winw1: [preflight] WARNING: Couldn't create the interface used for talking to the container runtime: crictl is required for container runtime: exec: "crictl": executable file not found in %PATH%
`
